src/mlpack/core/data/scaler_methods/whitening.hpp

```diff
@@ -0,0 +1,147 @@
+ * Whitening means the covariance matrix of the result is the identity matrix.
+ * 
+ * For whitening related formula and more info, check the link below.
+ * http://ufldl.stanford.edu/tutorial/unsupervised/PCAWhitening/
+ *
+ * @code
+ * arma::mat input;
+ * Load("train.csv", input);
+ * arma::mat output;
+ *
+ * // Scale the features using PCA.
+ * Whitening scale;
+ * scale.PCA(input, output);
+ *
+ * // Scale the features using ZCA.
+ * scale.ZCA(input, ouput);
+ * @endcode
+ */
+class Whitening
+{
+ public:
+  /**
+  * A constructor to set the regulatization parameter.
+  *
+  * @param eps Regularization parameter.
+  */
+  Whitening(double eps = 0.00005)
+  {
+    epsilon = eps;
+  }
+
+  /**
+  * Function for PCA whitening.
+  *
+  * @param input Dataset to scale features.
+  * @param output Output matrix with whitened features.
+  */
+  template<typename MatType>
+  void PCA(const MatType& input, MatType& output)
+  {
+    output.copy_size(input);
+    itemMean = arma::mean(input, 1);
+    // Centering the Dataset
+    output = (input.each_col() - itemMean);
+    arma::mat diag;
+    // Get eigenvectors and eigenvalues of covariance of input matrix.
+    eig_sym(eigenValues, eigenVectors, mlpack::math::ColumnCovariance(output));
+    // Generate diagonal matrix using 1 / sqrt(eigenvalues + epsilon) for each 
+    // value.
+    for (size_t i = 0; i < eigenValues.n_elem; i++)
+        eigenValues(i) = eigenValues(i) + epsilon;
+    mlpack::math::VectorPower(eigenValues, -0.5);
+    diag.zeros(eigenValues.n_elem, eigenValues.n_elem);
+    diag.diag() = eigenValues;
+
+    // Our whitening matrix is diag(1 / sqrt(eigenvectors)) * eigenvalues.
+    whiteningMatrix = diag * trans(eigenVectors);
+
+    // Now apply the whitening matrix.
+    output = whiteningMatrix * output;
+  }
+
+  /**
+  * Function for ZCA whitening.
+  *
+  * @param input Dataset to scale features.
+  * @param output Output matrix with whitened features.
+  */
+  template<typename MatType>
+  void ZCA(const MatType& input, MatType& output)
+  {
+    output.copy_size(input);
+    itemMean = arma::mean(input, 1);
+    // Centering the Dataset
+    output = (input.each_col() - itemMean);
+    arma::mat diag;
+    // Get eigenvectors and eigenvalues of covariance of input matrix.
+    eig_sym(eigenValues, eigenVectors, mlpack::math::ColumnCovariance(output));
+    // Generate diagonal matrix using 1 / sqrt(eigenvalues + epsilon) for each 
+    // value.
+    for (size_t i = 0; i < eigenValues.n_elem; i++)
+        eigenValues(i) = eigenValues(i) + epsilon;
+    mlpack::math::VectorPower(eigenValues, -0.5);
+    diag.zeros(eigenValues.n_elem, eigenValues.n_elem);
+    diag.diag() = eigenValues;
+
+    // Our whitening matrix is diag(1 / sqrt(eigenvectors)) * eigenvalues.
+    whiteningMatrix = diag * trans(eigenVectors);
+
+    // Now apply the whitening matrix.
+    output = whiteningMatrix * output;
+    // ZCA output.
+    output = eigenVectors * output;
```

Comment:
There's a lot of code duplication here, but I think that you could use the PCA code, and then have an extra overload that returns the eigenvectors, then use that for ZCA.

Reply:
I agree, Tried to avoid duplication of code, but didn't use overload function, and instead copied the data.